src/oci/runc.rs

```diff
@@ -0,0 +1,299 @@
+//! Interface to [runc][0], a OCI compliant cli for spawning and running containers.
+//!
+//! [0]: https://github.com/opencontainers/runc
+
+
+#![allow(dead_code)]
+
+use anyhow::{Context, Result};
+use async_trait::async_trait;
+use derive_builder::Builder;
+use dyn_clone::{clone_trait_object, DynClone};
+use getset::{Getters, Setters};
+use std::{
+    fmt::{self, Debug},
+    path::{Path, PathBuf},
+    process::Output,
+    string::ToString,
+};
+use strum::AsRefStr;
+use tokio::process::Command;
+use std::ffi::OsStr;
+
+const RUNC_SUB_CMD_CHECKPOINT : &str = "checkpoint";
+const RUNC_SUB_CMD_CREATE : &str = "create";
+const RUNC_SUB_CMD_DELETE : &str = "delete";
+const RUNC_SUB_CMD_EVENTS : &str = "events";
+const RUNC_SUB_CMD_EXEC : &str = "exec";
+const RUNC_SUB_CMD_INIT : &str = "init";
+const RUNC_SUB_CMD_KILL : &str = "kill";
+const RUNC_SUB_CMD_LIST : &str = "list";
+const RUNC_SUB_CMD_PAUSE : &str = "pause";
+const RUNC_SUB_CMD_PS : &str = "ps";
+const RUNC_SUB_CMD_RESTORE : &str = "restore";
+const RUNC_SUB_CMD_RESUME : &str = "resume";
+const RUNC_SUB_CMD_RUN : &str = "run";
+const RUNC_SUB_CMD_SPEC : &str = "spec";
+const RUNC_SUB_CMD_START : &str = "start";
+const RUNC_SUB_CMD_STATE : &str = "state";
+const RUNC_SUB_CMD_UPDATE : &str = "update";
+
+#[derive(Builder, Debug, Getters, Setters)]
+#[builder(pattern = "owned", setter(into))]
+// Runc is the main structure to be used when interacting with the container monitor.
+pub struct Runc {
+    #[getset(get, set)]
+    #[builder(private, default = "Box::new(DefaultRuncExecCommand)")]
+    /// The executor for runc
+    exec: Box<dyn ExecCommand>,
+
+    #[get]
+    /// Path to the runc binary
+    binary: PathBuf,
+}
+
+impl Runc {
+    /// Run 'runc checkpoint' with the provided args and return the output if the command execution succeeds.
+    /// This can still mean that runc itself failed, which can be verified via the exist status
+    /// of the output.
+    pub async fn checkpoint(&self, args: &[RuncCheckpointArgs]) -> Result<Output> {
+        self.exec().run_output(self.binary(), OsStr::new(RUNC_SUB_CMD_CHECKPOINT), args.iter().map(ToString::to_string).collect()).await
+    }
+
+    /// Run 'runc create' with the provided args and return the output if the command execution succeeds.
+    /// This can still mean that runc itself failed, which can be verified via the exist status
+    /// of the output.
+    pub async fn create(&self, args: &[RuncGlobalArgs]) -> Result<Output> {
```

Comment:
The idea with the functions here was to give feedback and appropriate flags for each sub command in runc making it easier to use. the problem is it's a lot of repeated code, Was thinking maybe we could do this with an enum instead but i'm not sure if i'm off on a crazy path with that idea (not accustom to using enums like that :) )

Reply:
Yes an enum seems appropriate in this place. :+1: